File: src/main/java/komu/blunt/core/CoreSequenceExpression.java
package komu.blunt.core;

import komu.blunt.asm.Assembler;
import komu.blunt.asm.Instructions;
import komu.blunt.asm.Linkage;
import komu.blunt.asm.Register;

import java.util.ArrayList;
import java.util.List;

import static java.util.Arrays.asList;

public final class CoreSequenceExpression extends CoreExpression {

    private final List<CoreExpression> expressions;
    
    public CoreSequenceExpression(List<CoreExpression> expressions) {
        this.expressions = new ArrayList<>(expressions);
    }

    public CoreSequenceExpression(CoreExpression... expressions) {
        this.expressions = asList(expressions);
    }

    @Override
    public Instructions assemble(Assembler asm, Register target, Linkage linkage) {
        Instructions instructions = new Instructions();

        for (CoreExpression exp : allButLast())
            instructions.append(exp.assemble(asm, target, Linkage.NEXT));

        instructions.append(last().assemble(asm, target, linkage));

        return instructions;
    }

    private CoreExpression last() {
        return expressions.get(expressions.size()-1);
    }

    private List<CoreExpression> allButLast() {
        return expressions.subList(0, expressions.size()-1);
    }

    @Override
    public String toString() {
        return expressions.toString();
    }
}

File: src/main/kotlin/komu/blunt/core/CoreSequenceExpression.kt
package komu.blunt.core

import komu.blunt.asm.*
import komu.blunt.utils.init
import java.util.*

class CoreSequenceExpression (private val expressions: List<CoreExpression>) : CoreExpression() {

    override fun assemble(asm: Assembler, target: Register, linkage: Linkage) =
        instructions {
            if (expressions.any()) {
                for (exp in expressions.init)
                    instructionsOf(exp.assemble(asm, target, Linkage.NEXT))

                instructionsOf(expressions.last().assemble(asm, target, linkage))
            } else {
                finishWithLinkage(linkage)
            }
        }

    override fun toString() = "(begin $expressions)"

    override fun simplify(): CoreExpression {
        val exps = ArrayList<CoreExpression>()
        for (exp in expressions) {
            val sexp = exp.simplify()
            if (sexp is CoreSequenceExpression)
                exps.addAll(sexp.expressions)
            else
                exps.add(sexp)
        }

        return CoreExpression.sequence(exps)
    }
}
